exe/tests/helpers/mod.rs

```diff
@@ -1,19 +1,87 @@
 use std::ffi::OsStr;
-use std::io;
+use std::fs;
 use std::path::Path;
+use std::path::PathBuf;
 use std::process::Command;
 use std::str;
 
 use anyhow::bail;
 use assert_cmd::cargo::CommandCargoExt;
+use std::env;
 
-pub fn get_image<P: AsRef<Path>>(to_dir: P) -> io::Result<()> {
+pub fn get_image<P: AsRef<Path>>(to_dir: P) -> anyhow::Result<PathBuf> {
     let image = "ubuntu";
     let tag = "latest";
     if to_dir.as_ref().exists() {
-        return Ok(());
+        let rootfs = to_dir.as_ref().join("rootfs");
+        if rootfs.exists() {
+            return Ok(rootfs);
+        } else {
+            bail!(
+                "{:?} exists but does not have a rootfs directory in it",
+                to_dir.as_ref().display()
+            );
+        }
     }
-    docker_extract::extract_image(image, tag, to_dir.as_ref())
+
+    let mut xdg_data_home_default = env::var("HOME").unwrap();
+    xdg_data_home_default.push_str("/.local/share");
+
+    let mut xdg_data_home = env::var("XDG_DATA_HOME").unwrap_or(xdg_data_home_default.clone());
+    if xdg_data_home.is_empty() {
+        xdg_data_home = xdg_data_home_default;
+    }
+
+    let puzzlefs_data_path = format!("{xdg_data_home}/puzzlefs");
+    fs::create_dir_all(&puzzlefs_data_path)?;
+
+    // skopeo copy docker://ubuntu:latest oci:$HOME/.local/share/puzzlefs/ubuntu:latest
```

Comment:
This will work ok the way you have it, but may be confusing:  You're creating an OCI layout .local/share/puzzlefs/ubuntu , and in there are creating an image called "latest".

so there will be .local/share/puzzlefs/ubuntu/index.json with an entry called "latest".

This is proper zot format if .local/share/puzzlefs is the zot root.  But it's a bit unorthodox otherwise.

Anyway I'll leave this comment here, I'm not sure whether it's better to leave it as is or change it.

Reply:
What can I do to improve this?